## Step 1. download and install GMTK

The Graphical Models Toolkit (GMTK) is an open source, publicly available toolkit for rapidly prototyping statistical models using dynamic graphical models (DGMs) and dynamic Bayesian networks (DBNs). 
It can be download at its [website](http://melodi.ee.washington.edu/gmtk/). Please follow its instruction to download and install.

## Step 2. follow preprocessing steps and prepare input file 

We assume that bam files of the biopsies (and the corresponding index files) are available for analysis. 
Nine preprocessing steps are needed in order to generate input files for THEMIS.
One example data file is combinedData_01-1-B2_01-2-B2_01-4-B3.txt.

1. Identify germline heterozygous sites

2. Identify SNVs 

3. Generate bed files for heterozygous sites and homozygous sites

4. perform MuTect and extract read counts from bam files

5. process MuTect raw result  

6. generate wig file from bam file with HMMcopy 

7. correct gc content and mappability

8. Combine MuTect processed data and HMMcopy processed data

9. Generate data files for GMTK

## Step 3. set up model specification files (e.g. structure file and master files)

In the directory `./model/`, users have to provide some model specification files, as follows. 
Please refer to doc/doc.md for documentation for these files.

* `hmm_factorialModel.mtr`   master file which specifies the parameter of the DBN
* `hmm_factorialModel.str`   structure file which specifies the structure of the DBN
* `constants.inc`   constant used in the structure file and master file
* `initCovars.txt`   initial values for covariance in the Gaussians
* `initDpmfs.txt`   initial values for Gaussian mixtures
* `init_hmm_factorialModel.params`   initial values for other parameters
* `params.notrain`   which parameters are not to be estimated. If a parameter is specified in this file, it will be fixed at its initial value rather than being estimated

As a part of the training/testing script, some java code is called to generate files which define the Gaussian distributions used by THEMIS. 
Please refer to doc/doc.md for documentation for these files.

## Step 4. execute shell scripts for triangulation (tri.sh) and model training/testing (trainTest.sh)

In the directory `./bin/`, users have to execute the following two scripts.

`tri.sh` is the shell script for triangulation of the graph.

`trainTest.sh` contain all the script for training the model and testing the model. 

## Step 5. the output of THEMIS (i.e. Viterbi path) 

The outputs of THEMIS are in the directory `./out/`, including

* `hmm_factorialModel_vitVals.txt`  // Viterbi path inferred from GMTK, which contains the genotype at each genomic site and the clone in which the mutation occurs
* `curr.txt`  // screen print from GMTK
* `err.txt`  // error messages from GMTK
* `em_covar.txt`  // estimated parameters for transition of G and Z variables
* `jt_info.txt`  // junction tree, automatically generated by GMTK
* `ll_iter{k}.txt`  // log likelihood yielded in the k-th iteration  
* `trained_factorialModel_{k}.params`  // parameters estimated in the k-th iteration  


